### PR TITLE
Add autofill integration to the keyboard

### LINKED ISSUE #199: Support password manager integration
In recent versions of Android, Password managers can now integrate into the keyboard. I probably don't have to explain how that is useful :)

https://developer.android.com/guide/topics/text/ime-autofill

### PR DESCRIPTION
**Pull request by esensar**
* Activated autofill support for our input method in method.xml
* Implemented required methods in our IME
* Styled views as similar as possible to the clipboard item
* Added InlineContentViewHorizontalScrollView which properly clips these views, since they are owned by another process and are usually drawn above our views

This closes #199

### CODE REVIEW

**Comment by esensar** [comment]
I have recorded a simple demonstration of the implementation.

https://github.com/SimpleMobileTools/Simple-Keyboard/assets/2764831/920a001d-a9d8-4db4-9110-9d5e2ac9e9e1

There are some things that I haven't implemented, since I am not sure what UX we are aiming for exactly:
- Adding clipboard icon to the clipboard item for visibility
- Further adjustments to the min and max width (we can also go for just wrapping width of these views, but that may become too much)
- I have used the same text color for subtitle, but usually apps go for a bit lighter color, to ensure it is visible. I was not sure if we already had such a color as a part of our theme, so I just used the same for now.
- I am having issues with getting the exact same background as the clipboard item, but on the other hand, maybe it is better to have slight distinction

**Comment by tibbi** [comment]
how to reproduce this? What should I do to see the autofill values?

**Comment by esensar** [comment]
They should appear automatically for the last 2 lines in that included sample screen, but before that you need to have an autofill service configured in your system settings. In my case it is `Bitwarden` application, but there should be the default `Google` service I think. If you have it set up, something should pop up.

It is also possible that you have no values that can be autofilled. You can then manually add some value in your autofill service to test with this. With `Google` it is done through `Passwords & Accounts` in system settings. From there it is possible to add a new password and select this application for it.

File: app/src/main/kotlin/com/simplemobiletools/keyboard/views/MyKeyboardView.kt [hunk 1]
```diff
@@ -1614,4 +1620,25 @@ class MyKeyboardView @JvmOverloads constructor(context: Context, attrs: Attribut
...
+                mToolbarHolder?.autofill_suggestions_holder?.addView(it, 0)
+            } else {
+                mToolbarHolder?.autofill_suggestions_holder?.addView(it)
+            }
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.R)
+    fun clearClipboardViews() {
+        if (mToolbarHolder?.autofill_suggestions_holder != null) {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/keyboard/views/MyKeyboardView.kt:1640]: this check doesnt make sense, if we check nullability below anyway

File: app/src/main/kotlin/com/simplemobiletools/keyboard/views/InlineContentViewHorizontalScrollView.kt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        allViews.forEach {
+            if (it is InlineContentView) {
+                it.beVisible()
+            }
+        }
+    }
+
+    private fun clipDescendantInlineContentViews() {
+        // This is only needed for InlineContentViews which are not available before this version
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/keyboard/views/InlineContentViewHorizontalScrollView.kt:72]: lets use our helper functions like !isRPlus() etc

**Comment by tibbi** [comment]
lets center everything horizontally, even clipboard is moved at the side now

**Comment by esensar** [comment]
> lets center everything horizontally, even clipboard is moved at the side now

Should it look like before when there are no suggestions, and move it to the side like it is now only when we add autofill suggestions? Because I guess we would need to have some kind of an indication to the users that they have other suggestions other than the clipboard.

File: app/src/main/kotlin/com/simplemobiletools/keyboard/views/InlineContentViewHorizontalScrollView.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import android.view.ViewTreeObserver.OnDrawListener
+import android.widget.HorizontalScrollView
+import android.widget.inline.InlineContentView
+import androidx.annotation.AttrRes
+import androidx.core.view.allViews
+import com.simplemobiletools.commons.extensions.beInvisible
+import com.simplemobiletools.commons.extensions.beVisible
+import com.simplemobiletools.commons.helpers.isRPlus
+
+
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/keyboard/views/InlineContentViewHorizontalScrollView.kt:17]: avoid inserting 2 empty lines

**Comment by tibbi** [comment]
> > lets center everything horizontally, even clipboard is moved at the side now
> 
> Should it look like before when there are no suggestions, and move it to the side like it is now only when we add autofill suggestions? Because I guess we would need to have some kind of an indication to the users that they have other suggestions other than the clipboard.

oh, didnt know both can be shown at once too. Anyway, ye, lets definitely center the clipboard field, if autofill is empty. Once there is some autofilling, it can be shown as it is now. But autofill is a really rare thing, we shouldnt change the default behaviour because of it.

**Comment by tibbi** [comment]
the clips background shouldnt be that wide though, just wrap its content and be centered as a whole

clipboard

**Comment by esensar** [comment]
> the clips background shouldnt be that wide though, just wrap its content and be centered as a whole
> 
> clipboard

My bad, for some reason I thought that was the previous behavior, but looks like I always had long strings in clipboard :smile:

File: app/src/main/kotlin/com/simplemobiletools/keyboard/views/MyKeyboardView.kt [hunk 4]
```diff
@@ -1632,18 +1639,32 @@ class MyKeyboardView @JvmOverloads constructor(context: Context, attrs: Attribut
...
+    private fun updateSuggestionsToolbarLayout() {
+        if (hasInlineViews()) {
+            // make room on suggestion toolbar for inline views
+            mToolbarHolder?.suggestions_items_holder?.gravity = Gravity.NO_GRAVITY
+            mToolbarHolder?.clipboard_value?.maxWidth = resources.getDimensionPixelSize(R.dimen.suggestion_max_width)
+        } else {
+            // restore original clipboard toolbar appearance
+            mToolbarHolder?.suggestions_items_holder?.gravity = Gravity.CENTER_HORIZONTAL
+            mToolbarHolder?.suggestions_holder?.measuredWidth?.also { maxWidth ->
+                mToolbarHolder?.clipboard_value?.maxWidth = maxWidth
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/keyboard/views/MyKeyboardView.kt:1661]: guess we could make this part nicer with using mToolbarHolder?.apply {...} and wrap whole if/else in it too

**Comment by esensar** [comment]
> theres still too much padding on sides, use whatever values are on main branch

Right, I forgot I added a minimum width. It should be the same as main branch now.

**Comment by tibbi** [comment]
is that autofill loaded somehow asynchronously? If I click on some edittext, the clipboard snippet blinks a bit in the middle, then it moves to the left

**Comment by tibbi** [comment]
okay thanks, its not that big deal, will see what feedback we get

**Comment by tibbi** [comment]
actually, lets check it with light theme. While the clipboard item is styled well, the autofill one is still dark

**Comment by tibbi** [comment]
@esensar just pinging you to recheck light theme

**Pull request merged**